CLDR-16836 kbd: add EBNF to spec for transforms (#4261)
unicode-org/cldr: docs/ldml/tr35-keyboards.md

    Emits the named mark. Also see [Markers](#markers).

#### Transform Grammar

##### Transform From Grammar

The `from=` attribute MUST match the `from-match` rule in this grammar. Not all strings which match this grammar are valid, specifically

The following is the [LDML EBNF](./tr35.md#ebnf) format for the grammar:

```ebnf
[ wfc: No more than 9 capture groups may be present. ]
[ vc: all variables referenced must be defined in the <variables> element ]

> [ vc: The CLDR repository may define additional constraints on the repertoire, such as requiring all characters to be in a published Unicode version and disallowing private-use characters. ]

Reviewer: ```suggestion
[ vc: If a keyboard definitions is submitted to the the CLDR repository, it must satisfy additional constraints on the character repertoire. For more information, see [CLDR keyboard repertoire constraints](#repertoire-constraints). ]
```
We have to point to where those constraints are documented. So add a little section header for that and point to it. Also make a change to a similar line below.

I suggest that the contents of that section be:

- No characters can have any of the following General_Category values in the latest version of the Unicode Standard: 
    - Private Use (Co)
    - Surrogate (Cs)
    - Unassigned (Cn)

Author: Not sure we can link from ebnf to spec. 

Also the repository requirements are a separate activity - I don't see a reason to specify them here in detail (though I'll take the list back to kbd-wg). Perhaps it's better to just say- see a future version of the spec.

Author: What about this

```ebnf
[ vc: If a keyboard definitions is submitted to the the CLDR repository, it must satisfy additional constraints on the character repertoire. ]
```

Reviewer: The problem with that is there is no way for the reader to find out what those requirements are — we can't have undefined vc requirements.

Author: Then I'd propose to drop it from vc and just note.

> [ vc: The CLDR repository may define additional constraints on the repertoire, such as requiring all characters to be in a published Unicode version and disallowing private-use characters. ]

Author: ```suggestion
```

Reviewer: You can go ahead and commit these two suggestions for deleting.

LF       ::= #xA
CR       ::= #xD
HEXDIG   ::= DIGIT
           | 'A'
           | 'B'
           | 'C'
           | 'D'
           | 'E'
           | 'F'
LHEXDIG  ::= HEXDIG
           | 'a'
           | 'b'
           | 'c'
           | 'd'
           | 'e'
           | 'f'
NAMESTARTCHAR
         ::= [:_#xC0-#xD6#xD8-#xF6#xF8-#x2FF#x370-#x37D#x37F-#x1FFF#x200C-#x200D#x2070-#x218F#x2C00-#x2FEF#x3001-#xD7FF#xF900-#xFDCF#xFDF0-#xFFFD#x10000-#x10FFFF]
           | ALPHA
NAMECHAR ::= NAMESTARTCHAR
           | [-.#xB7#x300-#x36F#x203F-#x2040]
           | DIGIT
NMTOKEN  ::= NAMECHAR+
```

##### Transform To Grammar

This is the grammar for the `<transform to="…"/>` attribute.  The `to=` attribute MUST match the `to-replacement` rule in this grammar. Not all strings which match this grammar are valid:

The following is the [LDML EBNF](./tr35.md#ebnf) format for the grammar:

```ebnf
[ vc: A referenced capture group must be present in the from= match string. ]
[ vc: The `$[1:…]` set format may only be used where there is exactly one capture group with a set variable on the from= match string. ]
[ vc: all variables referenced must be defined in the <variables> element ]

> [ vc: The CLDR repository may define additional constraints on the repertoire, such as requiring all characters to be in a published Unicode version and disallowing private-use characters. ]

Author: ```suggestion
```

Reviewer: Ok.

    Emits the named mark. Also see [Markers](#markers).

#### Transform Grammar

##### Transform From Grammar

The `from=` attribute MUST match the `from-match` rule in this grammar. Not all strings which match this grammar are valid:

* It is an error if there are more than 9 capture groups

> * Unicode codepoints and escaped characters may not be invalid or unpaired surrogates

Reviewer 2: too many negatives, did you mean "may be invalid"?

Reviewer: That's why I mentioned the w3c well-formedness / validity constraints. We should not talk about errors, but rather constraints:
```
[ wfc: No more than 9 capture groups are allowed. ]
[ wfc: Surrogate code points are not allowed. ]
```
Note: the 2nd one can also be expressed in the BNF.

Author: I have moved all of this  into wfc / vc

* The CLDR repository may define additional requirements on the repertoire, such as requiring all characters to be in a published Unicode version and disallowing private-use characters.
* It is an error to reference variables that are not defined.

The following is the [LDML EBNF](./tr35.md#ebnf) format for the grammar:

```ebnf

from-match
         ::= '^'? atoms
atoms    ::= atom ( '|'? atom )*
atom     ::= quark quantifier?
quark    ::= non-group
           | group
non-group
         ::= simple-matcher
           | codepointseq
           | variable
variable ::= string-variable
           | set-variable
string-variable
         ::= '${' var-id '}'
set-variable
         ::= '$[' var-id ']'

> group    ::= capturing-group

Reviewer 2: Is there a supposed ordering to the grammar? It feels like the idea is to use first, define next, on depth-first basis, but then why is this not followed by `var-id` and then `codepointseq`? I don't mind whether its define-then-use or use-then-define or alphabetical, but it would help if it was systematic.

Author: There isn't a specific order. I can put them in some better order.

           | non-capturing-group
quantifier
         ::= bounded-quantifier
           | '?'

> codepointseq

Reviewer 2: how is `codepointseq` different from `codepoint`?

Author: The spec allows for a space separated list of code points… However, where specifying a character class range start or end, only a single code point must be allowed, that's why there are two different rules. 

I'll clarify this.

         ::= '\' 'u' '{' cphexseq '}'
codepoint
         ::= '\' 'u' '{' cphexseq '}'
bounded-quantifier
         ::= '{' DIGIT ',' DIGIT '}'
non-capturing-group
         ::= '(' '?' ':' atoms ')'
capturing-group
         ::= '(' catoms ')'
catoms   ::= catom+
catom    ::= cquark quantifier?
cquark   ::= non-group
cphexseq ::= cphex ( ' ' cphex )*
cphex    ::= LHEXDIG ( LHEXDIG ( LHEXDIG ( LHEXDIG ( LHEXDIG LHEXDIG? )? )? )? )?
simple-matcher
         ::= text-char
           | class
           | '.'
           | match-marker
match-marker
         ::= '\m{.}'
           | match-named-marker
match-named-marker
         ::= '\m{' marker-id '}'
marker-id
         ::= NMTOKEN

> var-id   ::= IDCHAR+

Reviewer 2: why are variables limited to ASCII?

Author: I think that's in the spec

fixed-class
         ::= '\' fixed-class-char
fixed-class-char
         ::= 's'
           | 'S'
           | 't'
           | 'r'
           | 'n'
           | 'f'
           | 'v'
           | '\'
           | '$'
           | 'd'
           | 'w'
           | 'D'
           | 'W'
           | '0'
set-class
         ::= '[' set-negator set-members ']'
set-members
         ::= set-member+
set-member
         ::= text-char
           | char-range
           | match-marker
char-range
         ::= range-edge '-' range-edge
range-edge
         ::= codepoint
           | range-char
set-negator
         ::= '^'?
text-char
         ::= content-char
           | ws
           | escaped-char
           | '-'
           | ':'
range-char
         ::= content-char
           | ws
           | escaped-char
           | '.'
           | '|'
           | '{'
           | '}'
content-char
         ::= ASCII-CTRLS
           | ASCII-PUNCT
           | ALPHA
           | DIGIT
           | NON-ASCII
escaped-char
         ::= '\' ( '\' | '{' | '|' | '}' )
ws       ::= [ #x3000]
           | HTAB
           | CR
           | LF
IDCHAR   ::= ALPHA
           | DIGIT
           | '_'
ASCII-CTRLS

>          ::= [#x1-#x8#xB-#xC#xE-#x1F]

Reviewer 2: these are not allowed in XML 1.0
